File: backend/src/modules/property-management/svc.ts
import type { PagedResponse } from "../../shared/types/pagination.ts";
import {
  type Property,
  type PropertyChild,
  type PropertyParentDocumentRequirement,
  type PropertyUser,
  CONDITION_KEY,
  type PropertyChildDocumentRequirement,
  REQUIREMENT_TYPE,
  CHILD_CONDITION_KEY,
} from "./model.ts";
import { type IPropertyManagementRepo } from "./repo.ts";
import type { IIdentityClient } from "./client.ts";
import type {
  ParentConditionKeys,
  ChildConditionKeys,
} from "../../shared/types/identity.ts";
import type { DocumentType } from "../../shared/types/reporting.ts";
import { formatAggregateError } from "../../shared/util/error.ts";
import type { AsyncResponseType } from "../../shared/types/response.ts";

export interface IPropertyManagementSvc {
  getAllProperties(
    pageSize: number,
    pageNumber: number,
  ): AsyncResponseType<PagedResponse<Property>>;
  getPropertyUser(
    propertyId: string,
    userId: string,
  ): AsyncResponseType<PropertyUser>;
  getDocumentRequirementsForGivenPropertyParent(
    propertyId: string,
    userId: string,
  ): AsyncResponseType<PropertyParentDocumentRequirement[]>;
  getDocumentRequirementsForGivenPropertyChild(
    propertyId: string,
    childId: string,
  ): AsyncResponseType<PropertyChildDocumentRequirement[]>;
  getAllPropertyChildrenForGivenParent(
    propertyId: string,
    parentId: string,
  ): AsyncResponseType<PropertyChild[]>;
  getAllPropertyChildren(
    propertyId: string,
  ): AsyncResponseType<PropertyChild[]>;
  incrementPropertyChildrenPointsForGivenParent(
    propertyId: string,
    parentId: string,
    childrenIds: string[],
    pointValue: number,
  ): AsyncResponseType<PropertyChild[]>;
  getPointValueForGivenPropertyParentDocumentByDocumentType(
    propertyId: string,
    documentType: DocumentType,
  ): AsyncResponseType<number>;
  getAllPropertyChildrenPaged(
    propertyId: string,
    pageSize: number,
    pageNumber: number,
  ): AsyncResponseType<PagedResponse<PropertyChild>>;
}

export class PropertyManagementSvc implements IPropertyManagementSvc {
  private repo: IPropertyManagementRepo;
  private identityClient: IIdentityClient;
  constructor(repo: IPropertyManagementRepo, identityClient: IIdentityClient) {
    this.identityClient = identityClient;
    this.repo = repo;
  }
  async getAllProperties(
    pageSize: number,
    pageNumber: number,
  ): AsyncResponseType<PagedResponse<Property>> {
    return this.repo.getAllProperties(pageSize, pageNumber);
  }
  async getPropertyUser(
    propertyId: string,
    userId: string,
  ): AsyncResponseType<PropertyUser> {
    return this.repo.getPropertyUser(propertyId, userId);
  }
  async getDocumentRequirementsForGivenPropertyParent(
    propertyId: string,
    userId: string,
  ): AsyncResponseType<PropertyParentDocumentRequirement[]> {
    const promiseResults = await Promise.all([
      this.repo.getAllPropertyParentDocumentRequirements(propertyId),
      this.identityClient.getParentConditionKeys(userId),
    ]);
    const errors = promiseResults
      .map((result) => result.error)
      .filter((error) => error !== undefined);
    if (errors.length > 0)
      return {
        data: undefined,
        error: formatAggregateError(errors),
      };
    const [allReqPromiseResult, conditionKeyPromiseResult] = promiseResults;
    const activeReqs: PropertyParentDocumentRequirement[] = [];
    allReqPromiseResult.data?.forEach(
      (element: PropertyParentDocumentRequirement) => {
        if (
          conditionKeyPromiseResult.data &&
          isParentRequirementActive(conditionKeyPromiseResult.data, element)
        ) {
          activeReqs.push(element);
        }
      },
    );
    return { data: activeReqs, error: undefined };
  }

  async getAllPropertyChildren(
    propertyId: string,
  ): AsyncResponseType<PropertyChild[]> {
    return await this.repo.getAllPropertyChildren(propertyId);
  }

  async getAllPropertyChildrenForGivenParent(
    propertyId: string,
    parentId: string,
  ): AsyncResponseType<PropertyChild[]> {
    const taskResults = await Promise.all([
      this.getAllPropertyChildren(propertyId),
      this.identityClient.getAllParentChildren(parentId),
    ]);
    const errors = taskResults
      .map((result) => result.error)
      .filter((error) => error !== undefined);
    if (errors.length > 0)
      return {
        data: undefined,
        error: formatAggregateError(errors),
      };
    const [propChildrenResult, parentChildrenResult] = taskResults;
    const propChildren = propChildrenResult.data || [];
    const parentChildren = parentChildrenResult.data || [];
    const parentChildIds = new Set(parentChildren.map((pc) => pc.child_id));
    const matchingChildren = propChildren.filter((pc) =>
      parentChildIds.has(pc.child_id),
    );
    return { data: matchingChildren, error: undefined };
  }

  async incrementPropertyChildrenPointsForGivenParent(
    propertyId: string,
    parentId: string,
    childrenIds: string[],
    pointValue: number,
  ): AsyncResponseType<PropertyChild[]> {
    return await this.repo.incrementPropertyChildrenPointsForGivenParent(
      propertyId,
      parentId,
      childrenIds,
      pointValue,
    );
  }

  async getPointValueForGivenPropertyParentDocumentByDocumentType(
    propertyId: string,
    documentType: DocumentType,
  ): AsyncResponseType<number> {
    return await this.repo.getPointValueForGivenPropertyParentDocumentByDocumentType(
      propertyId,
      documentType,
    );
  }

  async getDocumentRequirementsForGivenPropertyChild(
    propertyId: string,
    childId: string,
  ): AsyncResponseType<PropertyChildDocumentRequirement[]> {
    const promiseResults = await Promise.all([
      this.repo.getAllPropertyChildrenDocumentRequirements(propertyId),
      this.identityClient.getChildConditionKeys(childId),
    ]);
    const errors = promiseResults
      .map((result) => result.error)
      .filter((error) => error !== undefined);
    if (errors.length > 0)
      return {
        data: undefined,
        error: formatAggregateError(errors),
      };
    const [allReqPromiseResult, conditionKeyPromiseResult] = promiseResults;
    const activeReqs: PropertyChildDocumentRequirement[] = [];
    allReqPromiseResult.data?.forEach(
      (element: PropertyChildDocumentRequirement) => {
        if (
          conditionKeyPromiseResult.data &&
          isChildRequirementActive(conditionKeyPromiseResult.data, element)
        ) {
          activeReqs.push(element);
        }
      },
    );
    return { data: activeReqs, error: undefined };
  }

  async getAllPropertyChildrenPaged(
    propertyId: string,
    pageSize: number,
    pageNumber: number,
  ): AsyncResponseType<PagedResponse<PropertyChild>> {
    return await this.repo.getAllPropertyChildrenPaged(
      propertyId,
      pageSize,
      pageNumber,
    );
  }
}

function isChildRequirementActive(
  cK: ChildConditionKeys,
  r: PropertyChildDocumentRequirement,
): boolean {
  if (r.requirement_type == REQUIREMENT_TYPE.Always) {
    return true;
  }

  if (r.requirement_type == REQUIREMENT_TYPE.Conditional) {
    switch (r.condition_key) {
      case CHILD_CONDITION_KEY.HasDisability:
        return !!cK.has_disability;
      // TODO
      default:
        return false;
    }
  }
  return false;
}

function isParentRequirementActive(
  cK: ParentConditionKeys,
  r: PropertyParentDocumentRequirement,
): boolean {
  if (r.requirement_type == REQUIREMENT_TYPE.Always) {
    return true;
  }

  if (r.requirement_type == REQUIREMENT_TYPE.Conditional) {
    switch (r.condition_key) {
      case CONDITION_KEY.IsEmployed:
        return !!cK.is_employed;
      case CONDITION_KEY.IsSelfEmployed:
        return !!cK.is_self_employed;
      case CONDITION_KEY.IsStudent:
        return !!cK.is_student;
      case CONDITION_KEY.FiledTaxInDesiredLocation:
        return !!cK.filed_tax_in_desired_location;
      case CONDITION_KEY.ResidesInDesiredLocation:
        return !!cK.resides_in_desired_location;
      default:
        return false;
    }
  }
  return false;
}

File: apps/backend/src/modules/property_management/svc.py
import asyncio
from abc import ABC, abstractmethod
from typing import List
from uuid import UUID
from src.modules.property_management.client import IIdentityClient
from src.shared.types.modules.property_management.enum import (
    CHILD_CONDITION_KEY,
    CONDITION_KEY,
    REQUIREMENT_TYPE,
)
from src.shared.types.modules.property_management.model import (
    Property,
    PropertyChild,
    PropertyChildDocumentRequirement,
    PropertyParentDocumentRequirement,
)
from src.modules.property_management.repo import IPropertyManagementRepo
from src.shared.types.modules.identity.model import (
    ChildConditionKeys,
    ParentConditionKeys,
)
from src.shared.types.modules.reporting.enum import CHILD_DOCUMENT_TYPE, DOCUMENT_TYPE
from src.shared.types.pagination import PagedResponse


class IPropertyManagementSvc(ABC):
    @abstractmethod
    async def get_all_properties(
        self,
        page_size: int,
        page_number: int,
    ) -> PagedResponse[Property]:
        pass

    @abstractmethod
    async def get_document_requirements_for_given_property_parent(
        self, property_id: str, user_id: str
    ) -> List[PropertyParentDocumentRequirement]:
        pass

    @abstractmethod
    async def get_all_property_children_for_given_parent(
        self, property_id: str, parent_id: str
    ) -> List[PropertyChild]:
        pass

    @abstractmethod
    async def get_all_property_children(
        self,
        property_id: str,
    ) -> List[PropertyChild]:
        pass

    @abstractmethod
    async def get_document_requirements_for_given_property_child(
        self,
        property_id: str,
        child_id: str,
    ) -> List[PropertyChildDocumentRequirement]:
        pass

    @abstractmethod
    async def get_all_property_children_paged(
        self,
        property_id: str,
        page_size: int,
        page_number: int,
    ) -> PagedResponse[PropertyChild]:
        pass

    @abstractmethod
    async def get_point_value_for_given_property_parent_document_by_document_type(
        self,
        property_id: UUID,
        document_type: DOCUMENT_TYPE,
    ) -> int:
        pass

    @abstractmethod
    async def get_point_value_for_given_property_child_document_by_document_type(
        self,
        property_id: UUID,
        document_type: CHILD_DOCUMENT_TYPE,
    ) -> int:
        pass

    @abstractmethod
    async def increment_property_children_points_for_given_parent(
        self,
        property_id: UUID,
        children_ids: List[UUID],
        point_value: int,
    ) -> List[PropertyChild]:
        pass

    @abstractmethod
    async def get_property_child_by_id(
        self,
        property_id: UUID,
        child_id: UUID,
    ) -> PropertyChild:
        pass

    @abstractmethod
    async def get_document_requirements_for_given_property_parent_partner(
        self, property_id: UUID, partner_id: UUID
    ) -> List[PropertyParentDocumentRequirement]:
        pass


class PropertyManagementSvc(IPropertyManagementSvc):
    def __init__(
        self, repo: IPropertyManagementRepo, identity_client: IIdentityClient
    ) -> None:
        self.repo = repo
        self.identity_client = identity_client

    async def get_all_properties(
        self,
        page_size: int,
        page_number: int,
    ) -> PagedResponse[Property]:
        return await self.repo.get_all_properties(
            page_size=page_size, page_number=page_number
        )

    async def get_all_property_children(
        self,
        property_id: str,
    ) -> List[PropertyChild]:
        return await self.repo.get_all_property_children(property_id=property_id)

    async def get_document_requirements_for_given_property_parent(
        self, property_id: str, user_id: str
    ) -> List[PropertyParentDocumentRequirement]:
        async with asyncio.TaskGroup() as tg:
            all_req_task = tg.create_task(
                self.repo.get_all_property_parent_document_requirements(
                    property_id=property_id
                )
            )
            condition_keys_task = tg.create_task(
                self.identity_client.get_parent_condition_keys(user_id=user_id)
            )
        all_req = all_req_task.result()
        condition_keys = condition_keys_task.result()
        active_reqs: List[PropertyParentDocumentRequirement] = []
        for req in all_req:
            if self._is_parent_requirement_active(condition_keys, req):
                active_reqs.append(req)
        return active_reqs

    async def get_all_property_children_for_given_parent(
        self, property_id: str, parent_id: str
    ) -> List[PropertyChild]:
        async with asyncio.TaskGroup() as tg:
            prop_children_task = tg.create_task(
                self.get_all_property_children(property_id=property_id)
            )
            parent_children_task = tg.create_task(
                self.identity_client.get_all_parent_children(parent_id=parent_id)
            )
        prop_children = prop_children_task.result()
        parent_children = parent_children_task.result()
        parent_child_ids = {x.child_id for x in parent_children}
        return [pc for pc in prop_children if pc.child_id in parent_child_ids]

    async def get_document_requirements_for_given_property_child(
        self,
        property_id: str,
        child_id: str,
    ) -> List[PropertyChildDocumentRequirement]:
        async with asyncio.TaskGroup() as tg:
            all_req_task = tg.create_task(
                self.repo.get_all_property_children_document_requirements(property_id)
            )
            condition_key_task = tg.create_task(
                self.identity_client.get_child_condition_keys(child_id)
            )
        all_req = all_req_task.result()
        condition_keys = condition_key_task.result()
        active_reqs: List[PropertyChildDocumentRequirement] = []
        for req in all_req:
            if self._is_child_requirement_active(condition_keys, req):
                active_reqs.append(req)
        return active_reqs

    async def get_all_property_children_paged(
        self,
        property_id: str,
        page_size: int,
        page_number: int,
    ) -> PagedResponse[PropertyChild]:
        return await self.repo.get_all_property_children_paged(
            property_id=property_id, page_size=page_size, page_number=page_number
        )

    async def get_point_value_for_given_property_parent_document_by_document_type(
        self,
        property_id: UUID,
        document_type: DOCUMENT_TYPE,
    ) -> int:
        return await self.repo.get_point_value_for_given_property_parent_document_by_document_type(
            property_id=property_id,
            document_type=document_type,
        )

    async def increment_property_children_points_for_given_parent(
        self,
        property_id: UUID,
        children_ids: List[UUID],
        point_value: int,
    ) -> List[PropertyChild]:
        return await self.repo.increment_property_children_points_for_given_parent(
            property_id=property_id, point_value=point_value, children_ids=children_ids
        )

    async def get_point_value_for_given_property_child_document_by_document_type(
        self,
        property_id: UUID,
        document_type: CHILD_DOCUMENT_TYPE,
    ) -> int:
        return await self.repo.get_point_value_for_given_property_child_document_by_document_type(
            property_id=property_id,
            document_type=document_type,
        )

    async def get_property_child_by_id(
        self,
        property_id: UUID,
        child_id: UUID,
    ) -> PropertyChild:
        return await self.repo.get_property_child_by_id(
            property_id=property_id, child_id=child_id
        )

    async def get_document_requirements_for_given_property_parent_partner(
        self, property_id: UUID, partner_id: UUID
    ) -> List[PropertyParentDocumentRequirement]:
        async with asyncio.TaskGroup() as tg:
            all_req_task = tg.create_task(
                self.repo.get_all_property_parent_document_requirements(
                    property_id=property_id
                )
            )
            condition_keys_task = tg.create_task(
                self.identity_client.get_parent_partner_condition_keys(
                    partner_id=partner_id
                )
            )
        all_req = all_req_task.result()
        condition_keys = condition_keys_task.result()
        active_reqs: List[PropertyParentDocumentRequirement] = []
        for req in all_req:
            if self._is_parent_requirement_active(condition_keys, req):
                active_reqs.append(req)
        return active_reqs

    def _is_parent_requirement_active(
        self,
        cK: ParentConditionKeys,
        r: PropertyParentDocumentRequirement,
    ) -> bool:
        if r.requirement_type == REQUIREMENT_TYPE.always:
            return True

        if r.requirement_type == REQUIREMENT_TYPE.conditional:
            match r.condition_key:
                case CONDITION_KEY.is_employed:
                    return cK.is_employed
                case CONDITION_KEY.is_self_employed:
                    return cK.is_self_employed
                case CONDITION_KEY.is_student:
                    return cK.is_student
                case CONDITION_KEY.filed_tax_in_desired_location:
                    return cK.filed_tax_in_desired_location
                case CONDITION_KEY.resides_in_desired_location:
                    return cK.resides_in_desired_location
                case _:
                    return False
        return False

    def _is_child_requirement_active(
        self,
        cK: ChildConditionKeys,
        r: PropertyChildDocumentRequirement,
    ) -> bool:
        if r.requirement_type == REQUIREMENT_TYPE.always:
            return True

        if r.requirement_type == REQUIREMENT_TYPE.conditional:
            match r.condition_key:
                case CHILD_CONDITION_KEY.has_disability:
                    return cK.has_disability
                case CHILD_CONDITION_KEY.is_from_single_parent_family:
                    return cK.is_from_single_parent_family
                case _:
                    return False
        return False
